Enable test execution using a RESTful API

Create interface to execute CTL test-suite using RESTful API.

Reference: https://github.com/opengeospatial/ets-wms13/issues/34

The wcs20 has its own local resources(schema's) and the parser is trying to load schema from local resource but it failed with following error:

------------------------------------------------
Caused by: java.lang.RuntimeException: Failed to parse resource from http://ows.eox.at/cite/mapserver?SERVICE=WCS&VERSION=2.0.1&REQUEST=GetCapabilities 
 schema_reference.4: Failed to read schema document 'file:/C:/CATALINA_BASE/bin/file:/C:/TE_BASE/resources/lib/ets-wcs20-1.15-SNAPSHOT.jar!/wcs/2.0/wcsAll.xsd', because 1) could not find the document; 2) the document could not be read; 3) the root element of the document is not <xsd:schema>.
	at com.occamlab.te.parsers.XMLValidatingParser.parse(XMLValidatingParser.java:227)
	... 102 more

------------------------------------

Can you please provide me some pointers? why the class loader is referring the two paths i.e. catalina_base/bin and te_base/resource? Do you have any idea about it?

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
   <artifactId>ets-wcs20</artifactId>
   <groupId>org.opengis.cite</groupId>
   <version>1.15-SNAPSHOT</version>
-  <packaging>pom</packaging>
+  <!-- <packaging>pom</packaging> -->
 
   <name>OGC Web Coverage Service 2.0.1 - Executable Test Suite</name>
   <description>
@@ -59,8 +59,23 @@
     <ets-code>wcs</ets-code>
     <spec-version>2.0.1</spec-version>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <teamengine.version>4.0.5</teamengine.version>
-  </properties>  
+    <teamengine.version>5.3</teamengine.version>
+  </properties>
+  
+  <dependencies>
+    <dependency>
+      <groupId>org.opengis.cite.teamengine</groupId>
+      <artifactId>teamengine-spi-ctl</artifactId>
+      <version>${teamengine.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.12</version>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+    
   <reporting>
     <plugins>
       <plugin>
@@ -82,14 +97,49 @@
   <build>
     <plugins>
       <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.6.0</version>
+        <configuration>
+          <source>1.8</source>
+          <target>1.8</target>
+          <debug>false</debug>
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-jar-plugin</artifactId>
-        <version>2.5</version>
+        <version>3.0.2</version>
+        <configuration>
+          <archive>
+            <manifest>
+              <addDefaultImplementationEntries>true</addDefaultImplementationEntries>
+            </manifest>
+            <manifestEntries>
+              <Build-Host>${hostname}</Build-Host>
+              <Build-Time>${maven.build.timestamp}</Build-Time>
+              <Build-Revision>${buildNumber}</Build-Revision>
+            </manifestEntries>
+          </archive>
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>buildnumber-maven-plugin</artifactId>
+        <version>1.4</version>
         <executions>
           <execution>
-            <id>default-jar</id>
-            <phase>never</phase>
+            <phase>validate</phase>
+            <goals>
+              <goal>create</goal>
+            </goals>
           </execution>
         </executions>
+        <configuration>
+          <doCheck>false</doCheck>
+          <doUpdate>false</doUpdate>
+          <shortRevisionLength>10</shortRevisionLength>
+        </configuration>
       </plugin>
       <plugin>
         <artifactId>maven-surefire-plugin</artifactId>
@@ -100,10 +150,17 @@
       </plugin>
       <plugin>
         <artifactId>maven-assembly-plugin</artifactId>
-        <version>2.5</version>
+        <version>2.6</version>
         <configuration>
+          <archive>
+            <manifest>
+              <mainClass>org.opengis.cite.wcs20.CtlControllerr</mainClass>
+            </manifest>
+          </archive>
           <descriptors>
             <descriptor>${basedir}/src/assembly/ctl.xml</descriptor>
+            <descriptor>${basedir}/src/assembly/aio.xml</descriptor>
+            <descriptor>${basedir}/src/assembly/deps.xml</descriptor>
           </descriptors>
         </configuration>
         <executions>
@@ -158,6 +215,18 @@
         </configuration>
       </plugin>
     </plugins>
+    
+    <resources>
+      <resource>
+        <directory>${basedir}/src/main/resources</directory>
+        <filtering>true</filtering>
+      </resource>
+      <resource>
+        <directory>${basedir}/src/main/scripts</directory>
+        <targetPath>${ets-code}/${spec-version}</targetPath>
+      </resource>
+    </resources>
+    
   </build>
 
   <profiles>

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,6 @@
   <artifactId>ets-wcs20</artifactId>
   <groupId>org.opengis.cite</groupId>
   <version>1.15-SNAPSHOT</version>
-  <!-- <packaging>pom</packaging> -->
 
   <name>OGC Web Coverage Service 2.0.1 - Executable Test Suite</name>
   <description>
@@ -154,7 +153,7 @@
         <configuration>
           <archive>
             <manifest>
-              <mainClass>org.opengis.cite.wcs20.CtlControllerr</mainClass>
+              <mainClass>org.opengis.cite.wcs20.CtlController</mainClass>
             </manifest>
           </archive>
           <descriptors>
@@ -183,6 +182,7 @@
         </configuration>
       </plugin>
       <plugin>     
+        <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-site-plugin</artifactId>
         <version>3.4</version>
         <configuration>
@@ -224,6 +224,11 @@
       <resource>
         <directory>${basedir}/src/main/scripts</directory>
         <targetPath>${ets-code}/${spec-version}</targetPath>
+      </resource>
+      <resource>
+        <directory>src/main/javadoc</directory>
+        <filtering>true</filtering>
+        <targetPath>doc/${ets-code}/${project.version}</targetPath>
       </resource>
     </resources>
     

--- a/src/assembly/aio.xml
+++ b/src/assembly/aio.xml
@@ -1,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<assembly xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.3">
+  <!-- Generate all-in-one jar including dependencies -->
+  <id>aio</id>
+  <formats>
+    <format>jar</format>
+  </formats>
+  <includeBaseDirectory>false</includeBaseDirectory>
+  <dependencySets>
+    <dependencySet>
+      <outputDirectory>/</outputDirectory>
+      <useProjectArtifact>true</useProjectArtifact>
+      <unpack>true</unpack>
+      <scope>runtime</scope>
+      <excludes>
+        <exclude>com.sun.jersey:jersey-server</exclude>
+      </excludes>
+    </dependencySet>
+  </dependencySets>
+</assembly>

--- a/src/assembly/deps.xml
+++ b/src/assembly/deps.xml
@@ -1,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
+<assembly xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2 
+	                    http://maven.apache.org/xsd/assembly-1.1.2.xsd">
+  <id>deps</id>
+  <formats>
+    <format>zip</format>
+    <format>tar.gz</format>
+  </formats>
+  <includeBaseDirectory>false</includeBaseDirectory>
+  <dependencySets>
+    <dependencySet>
+      <useProjectArtifact>true</useProjectArtifact>
+      <excludes>
+        <exclude>org.opengis.cite.teamengine:teamengine-spi-ctl</exclude>
+      </excludes>
+      <outputDirectory>/</outputDirectory>
+      <unpack>false</unpack>
+      <scope>runtime</scope>
+      <useTransitiveDependencies>true</useTransitiveDependencies>
+      <useTransitiveFiltering>true</useTransitiveFiltering>
+    </dependencySet>
+  </dependencySets>
+</assembly>

--- a/src/main/config/config.xml
+++ b/src/main/config/config.xml
@@ -9,7 +9,7 @@
       <suite>
         <namespace-uri>http://www.opengis.net/wcs/2.0</namespace-uri>
         <prefix>wcs2</prefix>
-        <local-name>suite</local-name>
+        <local-name>ets-wcs20</local-name>
         <title>WCS 2.0 Conformance Test Suite</title>
         <description>Checks WCS Core implementations for conformance with OGC 09-110r3 and 09-110r4.</description>
       </suite>

--- a/src/main/java/org/opengis/cite/wcs20/CtlController.java
+++ b/src/main/java/org/opengis/cite/wcs20/CtlController.java
@@ -1,0 +1,125 @@
+package org.opengis.cite.wcs20;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.transform.Source;
+
+import org.w3c.dom.Document;
+
+import com.occamlab.te.SetupOptions;
+import com.occamlab.te.spi.ctl.CtlExecutor;
+import com.occamlab.te.spi.executors.TestRunExecutor;
+import com.occamlab.te.spi.jaxrs.TestSuiteController;
+
+public class CtlController implements TestSuiteController{
+	
+	private TestRunExecutor executor;
+    private Properties etsProperties = new Properties();
+    
+	public CtlController() {
+		SetupOptions setupOpts = new SetupOptions();
+        try (InputStream is = getClass().getResourceAsStream("ets.properties")) {
+            this.etsProperties.load(is);
+            String mainScriptPath = etsProperties.getProperty("main-script");
+            File ctlFile = findScriptFile(URI.create(mainScriptPath));
+            setupOpts.addSource(ctlFile);
+            this.executor = new CtlExecutor(setupOpts);
+        } catch (IOException ex) {
+            Logger.getLogger(getClass().getName()).log(Level.SEVERE, "Failed to initialize CtlController. {0}",
+                    ex.getMessage());
+        }
+	}
+	
+	/**
+     * A convenience method for running the test suite using a command-line
+     * interface.
+     *
+     * @param args
+     *            Test run arguments (optional). The first argument must refer
+     *            to a local XML properties file containing the expected set of
+     *            test run arguments. If no argument is supplied, the file
+     *            located at ${user.home}/test-run-props.xml will be used.
+     * @throws Exception
+     *             If an error errors during the test run.
+     */
+    public static void main(String[] args) throws Exception {
+        File propsFile;
+        if (args.length == 0) {
+            propsFile = new File(System.getProperty("user.home"), "test-run-props.xml");
+        } else {
+            String xmlProps = args[0];
+            propsFile = (xmlProps.startsWith("file:")) ? new File(URI.create(xmlProps)) : new File(xmlProps);
+        }
+        if (!propsFile.isFile()) {
+            throw new IllegalArgumentException("Test run arguments not found at " + propsFile);
+        }
+        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+        DocumentBuilder db = dbf.newDocumentBuilder();
+        Document testRunArgs = db.parse(propsFile);
+        TestSuiteController controller = new CtlController();
+        Source results = controller.doTestRun(testRunArgs);
+        Logger.getLogger(CtlController.class.getName()).log(Level.INFO, "Test results: {0}", results.getSystemId());
+    }
+
+    @Override
+	public String getCode() {
+		return etsProperties.getProperty("ets-code");
+	}
+
+    @Override
+    public String getVersion() {
+		return etsProperties.getProperty("ets-version");
+	}
+
+    @Override
+    public String getTitle() {
+		return etsProperties.getProperty("ets-title");
+	}
+
+    @Override
+    public Source doTestRun(Document testRunArgs) throws Exception {
+		Properties validArgs = TestRunArguments.validateArguments(testRunArgs);
+        // pass Properties directly instead?
+        return executor.execute(TestRunArguments.propertiesAsDocument(validArgs));
+	}
+	
+	/**
+     * Creates a File from the given URI reference. If the URI is not absolute,
+     * is is resolved against the location of the TE_BASE/scripts directory; if
+     * a file does not exist at this location the URI is assumed to be a
+     * classpath reference.
+     *
+     * @param uri
+     *            An absolute or relative URI.
+     * @return A File object, or null if one could not be created.
+     */
+    final File findScriptFile(URI uri) {
+        File ctlFile = null;
+        File baseDir = SetupOptions.getBaseConfigDirectory();
+        if (!uri.isAbsolute()) {
+            File scriptsDir = new File(baseDir, "scripts");
+            ctlFile = new File(scriptsDir, uri.getPath());
+        }
+        if (null == ctlFile || !ctlFile.isFile()) {
+            URL resource = getClass().getResource(uri.getPath());
+            try {
+                ctlFile = new File(resource.toURI());
+            } catch (URISyntaxException ex) {
+                Logger.getLogger(getClass().getName()).log(Level.INFO, "Invalid URI: {0}", ex);
+            }
+        }
+        Logger.getLogger(getClass().getName()).log(Level.CONFIG, "Created File object: {0}", ctlFile);
+        return ctlFile;
+    }
+
+}

--- a/src/main/java/org/opengis/cite/wcs20/TestRunArguments.java
+++ b/src/main/java/org/opengis/cite/wcs20/TestRunArguments.java
@@ -39,7 +39,11 @@ public class TestRunArguments {
         } catch (NullPointerException e2) {
             throw new IllegalArgumentException(String.format("Missing required argument: %s", WCS20.URL));
         }
-        if (null == args.getProperty(WCS20.CORE)) {
+        if (null == args.getProperty(WCS20.CORE) && null == args.getProperty(WCS20.EXT_CRS) 
+        		&& null == args.getProperty(WCS20.EXT_INTERPOLATION) && null == args.getProperty(WCS20.EXT_POST) 
+        		&& null == args.getProperty(WCS20.EXT_PROCESSING) && null == args.getProperty(WCS20.EXT_RSUBSETTING)
+        		&& null == args.getProperty(WCS20.EXT_SCALING) && null == args.getProperty(WCS20.PROFILE_EOWCS)) {
+        	
             args.setProperty(WCS20.CORE, "core");
         }
         return args;

--- a/src/main/java/org/opengis/cite/wcs20/TestRunArguments.java
+++ b/src/main/java/org/opengis/cite/wcs20/TestRunArguments.java
@@ -1,0 +1,70 @@
+package org.opengis.cite.wcs20;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NodeList;
+
+public class TestRunArguments {
+
+	public static Properties validateArguments(Document testRunArgs) {
+        NodeList entries = testRunArgs.getDocumentElement().getElementsByTagName("entry");
+        if (entries.getLength() == 0) {
+            throw new IllegalArgumentException("No test run arguments.");
+        }
+        
+        Properties args = new Properties();
+        for (int i = 0; i < entries.getLength(); i++) {
+            Element entry = (Element) entries.item(i);
+            args.setProperty(entry.getAttribute("key"), entry.getTextContent().trim());
+        }
+        String capabilitiesUrl = args.getProperty(WCS20.URL);
+        try {
+            URI uriRef = new URI(capabilitiesUrl);
+            if (!uriRef.isAbsolute()) {
+                throw new IllegalArgumentException(String.format("Not an absolute URI: %s", uriRef));
+            }
+        } catch (URISyntaxException e1) {
+            throw new IllegalArgumentException(String.format("Invalid URI reference: %s", capabilitiesUrl));
+        } catch (NullPointerException e2) {
+            throw new IllegalArgumentException(String.format("Missing required argument: %s", WCS20.URL));
+        }
+        if (null == args.getProperty(WCS20.CORE)) {
+            args.setProperty(WCS20.CORE, "core");
+        }
+        return args;
+    }
+	
+	/**
+     * Creates a document that contains a simple XML representation of the given
+     * set of properties.
+     * 
+     * @param props
+     *            A set of properties (key-value pairs).
+     * @return A Document node, wherein each property is represented by an
+     *         entry[@key] element.
+     */
+    public static Document propertiesAsDocument(Properties props) {
+        Document doc = null;
+        try {
+            ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+            props.storeToXML(outStream, "Test run arguments");
+            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
+            doc = builder.parse(new ByteArrayInputStream(outStream.toByteArray()));
+        } catch (Exception e) { // very unlikely
+            Logger.getLogger(TestRunArguments.class.getName()).log(Level.WARNING, "Failed to create Document node. {0}",
+                    e.getMessage());
+        }
+        return doc;
+    }
+}

--- a/src/main/java/org/opengis/cite/wcs20/WCS20.java
+++ b/src/main/java/org/opengis/cite/wcs20/WCS20.java
@@ -1,0 +1,36 @@
+package org.opengis.cite.wcs20;
+
+public class WCS20 {
+
+	public WCS20() {
+		
+	}
+	
+	/** Reference to the WCS capabilities document (URI). */
+	public static final String URL = "url";
+	
+	/** The <b>WCS Core</b> is a conformance class executed by default. */
+	public static final String CORE = "core";
+	
+	/** Run test for <b>WCS POST</b> if parameter is provided. */
+	public static final String EXT_POST = "ext_post";
+	
+	/** Run test for <b>WCS WCPS 2.0 (Processing)</b> if parameter is provided. */
+	public static final String EXT_PROCESSING = "ext_proc";
+	
+	/** Run test for <b>WCS Scaling 1.0</b> if parameter is provided. */
+	public static final String EXT_SCALING = "ext_scal";
+	
+	/** Run test for <b>WCS Interpolation 1.0</b>  if parameter is provided. */
+	public static final String EXT_INTERPOLATION = "ext_int";
+	
+	/** Run test for <b>WCS Range Subsetting 1.0</b> if parameter is provided. */
+	public static final String EXT_RSUBSETTING = "ext_rsub";
+	
+	/** Run test for <b>WCS-CRS-extension 1.0</b> if parameter is provided. */
+	public static final String EXT_CRS = "ext_crs";
+	
+	/** Run test for <b>WCS EO 1.0 (EOWCS)</b> if parameter is provided. */
+	public static final String PROFILE_EOWCS = "profile_eowcs";
+	
+}

--- a/src/main/javadoc/index.html
+++ b/src/main/javadoc/index.html
@@ -1,0 +1,10 @@
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta http-equiv="refresh" content="0; url=overview.html" />
+    <title />
+  </head>
+  <body>
+  </body>
+</html>

--- a/src/main/javadoc/overview.html
+++ b/src/main/javadoc/overview.html
@@ -1,0 +1,94 @@
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
+  <head>
+    <title>WCS 2.0 Conformance Test Suite</title>
+    <meta charset="UTF-8"/>
+    <style type="text/css">
+      body {
+        color: black;
+        background: white;
+        font-family: Georgia, serif;
+      }
+      h1, h2, h3 {
+        font-family: Verdana, sans-serif;
+        color: #000099;
+        text-align: left;
+      }
+      h1 {
+        font-size: 1.4em;
+        margin-top: 2em;
+      }
+      h2 {
+        font-size: 1.2em;
+      }
+      h3 {
+        font-size: 1em;
+        margin-bottom: 0.5em;
+      }
+      table {
+        margin-top: 0.5em;
+        margin-bottom: 1em;
+        border-collapse: collapse;
+      }
+      td, th {
+        padding: 0.25em;
+        border: 1px solid black;
+      }
+      caption, figcaption {
+        font-weight: bold;
+        text-align: left;
+        margin: 0.25em;
+      }
+      th {
+        font-weight: bold;
+        font-family: Verdana, sans-serif;
+        text-align: left;
+        vertical-align: top;
+        background: #eeeeee;
+        color: #000099;
+      }
+      dt {
+        margin-top: 0.5em;
+        font-weight: bold;
+      }</style>
+  </head>
+  <body>
+    <h1>WCS 2.0 Conformance Test Suite</h1>
+    <p>This Executable Test Suite (ETS) verifies the conformance of WCS 2.0 services. 
+    The services should be accessed through WCS GetCapabilities and DescribeCoverage using the HTTP KVP, POST, and/or SOAP protocol bindings, and GetCoverage using the HTTP KVP binding. 
+    The GetCoverage response should be provided as a GML coveage instance document..</p>
+
+    <p>The test run arguments are summarized in the following table. The <em>Obligation</em> descriptor 
+    can have the following values: M (mandatory), O (optional), or C (conditional). </p>
+
+    <table border="1" style="border-collapse: collapse;">
+      <caption>Test run arguments</caption>
+      <thead>
+        <tr>
+          <th>Name</th>
+          <th>Value domain</th>
+          <th>Obligation</th>
+          <th>Description</th>
+        </tr>
+      </thead>
+      <tbody>
+        <tr>
+          <td>url</td>
+          <td>URI</td>
+          <td>M</td>
+          <td>An absolute URI that refers to an XML representation of the service capabilities document.</td>
+        </tr>
+        <tr>
+          <td colspan="4">
+            <strong>Notes</strong>
+            <ol>
+              <li>Ampersand ('&amp;') characters appearing within query parameter values must be
+                percent-encoded as %26.</li>
+            </ol>
+          </td>
+        </tr>
+      </tbody>
+    </table>
+
+  </body>
+</html>

--- a/src/main/javadoc/overview.html
+++ b/src/main/javadoc/overview.html
@@ -79,11 +79,60 @@
           <td>An absolute URI that refers to an XML representation of the service capabilities document.</td>
         </tr>
         <tr>
+          <td>core</td>
+          <td>core</td>
+          <td>O, default</td>
+          <td>The WCS Core profile.</td>
+        </tr>
+        <tr>
+          <td>ext_post</td>
+          <td>post</td>
+          <td>O</td>
+          <td>The WCS POST profile</td>
+        </tr>
+        <tr>
+          <td>ext_proc</td>
+          <td>processing</td>
+          <td>O</td>
+          <td>The WCS WCPS 2.0 (Processing) profile</td>
+        </tr>
+        <tr>
+          <td>ext_scal</td>
+          <td>scaling</td>
+          <td>O</td>
+          <td>The WCS Scaling 1.0 profile</td>
+        </tr>
+        <tr>
+          <td>ext_int</td>
+          <td>interpolation</td>
+          <td>O</td>
+          <td>The WCS Interpolation 1.0 profile</td>
+        </tr>
+        <tr>
+          <td>ext_rsub</td>
+          <td>range subsetting</td>
+          <td>O</td>
+          <td>The WCS Range Subsetting 1.0 profile</td>
+        </tr>
+        <tr>
+          <td>ext_crs</td>
+          <td>crs</td>
+          <td>O</td>
+          <td>The WCS-CRS-extension 1.0 profile</td>
+        </tr>
+        <tr>
+          <td>profile_eowcs</td>
+          <td>eowcs</td>
+          <td>O</td>
+          <td>The WCS EO 1.0 (EOWCS) profile</td>
+        </tr>
+        <tr>
           <td colspan="4">
             <strong>Notes</strong>
             <ol>
+              <li>If the profile has not provided in query parameter then default WCS Core profile will be executed. </li>
               <li>Ampersand ('&amp;') characters appearing within query parameter values must be
-                percent-encoded as %26.</li>
+                percent-encoded as %26.</li>  
             </ol>
           </td>
         </tr>

--- a/src/main/resources/META-INF/services/com.occamlab.te.spi.jaxrs.TestSuiteController
+++ b/src/main/resources/META-INF/services/com.occamlab.te.spi.jaxrs.TestSuiteController
@@ -1,0 +1,1 @@
+org.opengis.cite.wcs20.CtlController

--- a/src/main/resources/org/opengis/cite/wcs20/ets.properties
+++ b/src/main/resources/org/opengis/cite/wcs20/ets.properties
@@ -1,0 +1,4 @@
+ets-title = ${project.name}
+ets-version = ${project.version}
+ets-code = ${ets-code}
+main-script = /${ets-code}/${spec-version}/ctl/wcs2suite-auto.xml

--- a/src/main/resources/org/opengis/cite/wcs20/test-run-props.xml
+++ b/src/main/resources/org/opengis/cite/wcs20/test-run-props.xml
@@ -1,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE properties SYSTEM "http://java.sun.com/dtd/properties.dtd">
+<properties version="1.0">
+    <comment>Test run arguments for ets-wcs20</comment>
+    <entry key="url"></entry>
+    <entry key="core">core</entry>
+    <entry key="ext_post"></entry>
+    <entry key="ext_proc"></entry>
+    <entry key="ext_scal"></entry>
+    <entry key="ext_int"></entry>
+    <entry key="ext_rsub"></entry>
+    <entry key="ext_crs"></entry>
+    <entry key="profile_eowcs"></entry>
+</properties>

--- a/src/main/scripts/ctl/wcs2suite-auto.xml
+++ b/src/main/scripts/ctl/wcs2suite-auto.xml
@@ -1,6 +1,179 @@
-<ctl:suite xmlns:ctl="http://www.occamlab.com/ctl" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:wcs2="http://www.opengis.net/wcs/2.0" name="wcs2:ets-wcs20-auto">
-	<ctl:title>WCS 2.0 Test Suite</ctl:title>
-	<ctl:description>Validates WCS 2.0 Implementations</ctl:description>
-	<ctl:link>docs/wcs/2.0/</ctl:link>
-	<ctl:starting-test>wcs2:main</ctl:starting-test>
-</ctl:suite>
+<?xml version="1.0" encoding="UTF-8"?>
+<ctl:package xmlns:ctl="http://www.occamlab.com/ctl" xmlns:xsl="http://www.w3.org/1999/XSL/Transform" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ows="http://www.opengis.net/ows/2.0" xmlns:gml="http://www.opengis.net/gml/3.2" xmlns:wcs="http://www.opengis.net/wcs/2.0" xmlns:ctlp="http://www.occamlab.com/te/parsers" xmlns:wcs2="http://www.opengis.net/wcs/2.0" xmlns:sch="http://purl.oclc.org/dsdl/schematron" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:wcseo="http://www.opengis.net/wcseo/1.0" xmlns:xi="http://www.w3.org/2001/XInclude">
+	<!-- NOTE: As there is no canonical practice among and within communities dealing with coverages to use one particular protocol none of the protocols specified in WCS extensions is mandatory. In practice, this may lead to a lack of interoperability between client and server implementations. In future, user communities might agree eventually to make a particular protocol binding mandatory. We use the GetCapabilities in HTTP Get KVP as an initially starting point-->
+	<xi:include href="core2main.xml"/>
+	<xi:include href="DataModel.xml"/>
+	<xi:include href="DescribeCoverage.xml"/>
+	<xi:include href="eowcs_Capabilities.xml" />
+	<xi:include href="eowcs_DescribeCoverage.xml" />
+	<xi:include href="eowcs_DescribeEOCoverageSet.xml"/>
+	<xi:include href="eowcs_EO-data-model.xml"/>
+	<xi:include href="eowcs_functions.xml"/>
+	<xi:include href="eowcs_get-kvp.xml"/>
+	<xi:include href="eowcs_GetCoverage.xml"/>
+	<xi:include href="eowcs_main.xml"/>
+	<xi:include href="eowcs_service-model.xml"/>
+	<xi:include href="eowcs_soap.xml"/>
+	<xi:include href="eowcs_wcs_extensions.xml"/>
+	<xi:include href="ext_crs.xml"/>
+	<xi:include href="ext_get-int.xml"/>
+	<xi:include href="ext_get-proc.xml"/>
+	<xi:include href="ext_get-rsub.xml"/>
+	<xi:include href="ext_get-scal.xml"/>
+	<xi:include href="ext_post-xml.xml"/>
+	<xi:include href="extension.xml"/>
+	<xi:include href="GetCapabilities.xml"/>
+	<xi:include href="GetCoverage.xml"/>
+	<xi:include href="OWSCommon.xml"/>
+	<xi:include href="RequestBase.xml"/>
+	<xi:include href="wcps.xml"/>
+	
+	<ctl:suite xmlns:ctl="http://www.occamlab.com/ctl" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:wcs2="http://www.opengis.net/wcs/2.0" name="wcs2:ets-wcs20-auto">
+		<ctl:title>WCS 2.0 Test Suite</ctl:title>
+		<ctl:description>Validates WCS 2.0 Implementations</ctl:description>
+		<ctl:link>docs/wcs/2.0/</ctl:link>
+		<ctl:starting-test>wcs2:main</ctl:starting-test>
+	</ctl:suite>
+	<ctl:test name="wcs2:main">
+		<ctl:param name="url"/>
+		<ctl:param name="core"/>
+		<ctl:param name="ext_post"/>
+		<ctl:param name="ext_proc"/>
+		<ctl:param name="ext_scal"/>
+		<ctl:param name="ext_int"/>
+		<ctl:param name="ext_rsub"/>
+		<ctl:param name="ext_crs"/>
+		<ctl:param name="profile_eowcs"/>
+		<ctl:assertion>Prune query parameters if present.</ctl:assertion>
+		<ctl:code>
+			<xsl:choose>
+				<xsl:when test="contains($url,'?')">
+						<xsl:if test="contains($core,'core')">
+							<ctl:call-test name="wcs2:main_test">
+								<ctl:with-param name="url" select="substring-before($url,'?')"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_post,'post')">
+							<ctl:call-test name="wcs2:post-xml-main">
+								<ctl:with-param name="url" select="substring-before($url,'?')"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_proc,'processing')">
+							<ctl:call-test name="wcs2:get-proc-main">
+								<ctl:with-param name="url" select="substring-before($url,'?')"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_scal,'scaling')">
+							<ctl:call-test name="wcs2:get-scal-main">
+								<ctl:with-param name="url" select="substring-before($url,'?')"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_int,'interpolation')">
+							<ctl:call-test name="wcs2:get-int-main">
+								<ctl:with-param name="url" select="substring-before($url,'?')"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_rsub,'range subsetting')">
+							<ctl:call-test name="wcs2:get-rsub-main">
+								<ctl:with-param name="url" select="substring-before($url,'?')"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_crs,'crs')">
+							<ctl:call-test name="wcs2:crs">
+								<ctl:with-param name="url" select="substring-before($url,'?')"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($profile_eowcs,'eowcs')">
+							<ctl:call-test name="wcseo:main">
+								<ctl:with-param name="url" select="substring-before($url,'?')"/>
+							</ctl:call-test>
+						</xsl:if>
+				</xsl:when>
+				<xsl:otherwise>
+						<xsl:if test="contains($core,'core')">
+							<ctl:call-test name="wcs2:main_test">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_post,'post')">
+							<ctl:call-test name="wcs2:post-xml-main">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_proc,'processing')">
+							<ctl:call-test name="wcs2:get-proc-main">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_scal,'scaling')">
+							<ctl:call-test name="wcs2:get-scal-main">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_int,'interpolation')">
+							<ctl:call-test name="wcs2:get-int-main">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_rsub,'range subsetting')">
+							<ctl:call-test name="wcs2:get-rsub-main">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($ext_crs,'crs')">
+							<ctl:call-test name="wcs2:crs">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:if>
+						<xsl:if test="contains($profile_eowcs,'eowcs')">
+							<ctl:call-test name="wcseo:main">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:if>
+				</xsl:otherwise>
+			</xsl:choose>
+		</ctl:code>
+	</ctl:test>
+	<ctl:test name="wcs2:main_test" isConformanceClass="true" isBasic="true">
+		<ctl:param name="url"/>
+		<ctl:assertion>Test the WCS 2.0.</ctl:assertion>
+		<ctl:code>
+			<xsl:variable name="Capabilities">
+				<ctl:request>
+					<ctl:url>
+						<xsl:value-of select="$url"/>
+					</ctl:url>
+					<ctl:method>get</ctl:method>
+					<ctl:param name="SERVICE">WCS</ctl:param>
+					<ctl:param name="VERSION">2.0.1</ctl:param>
+					<ctl:param name="REQUEST">GetCapabilities</ctl:param>
+					<ctlp:XMLValidatingParser>
+						<ctlp:schemas>
+							<ctlp:schema type="resource">wcs/2.0/wcsAll.xsd</ctlp:schema>
+						</ctlp:schemas>
+					</ctlp:XMLValidatingParser>
+				</ctl:request>
+			</xsl:variable>
+			<xsl:choose>
+				<xsl:when test="not(empty($Capabilities))">
+					<ctl:message>[MESSAGE] Test the core and extensions when they are available.</ctl:message>
+					<xsl:choose>
+						<xsl:when test="(contains(string-join(($Capabilities//*[local-name()='ServiceIdentification']/*[local-name()='Profile']),' '), 'http://www.opengis.net/spec/WCS_protocol-binding_soap/1.0') or contains(string-join(($Capabilities//*[local-name()='ServiceIdentification']/*[local-name()='Profile']),' '), 'http://www.opengis.net/spec/WCS_protocol-binding_post-xml/1.0') or contains(string-join(($Capabilities//*[local-name()='ServiceIdentification']/*[local-name()='Profile']),' '), 'http://www.opengis.net/spec/WCS_protocol-binding_get-kvp/1.0'))">
+							<ctl:call-test name="wcs2:core-main">
+								<ctl:with-param name="url" select="$url"/>
+							</ctl:call-test>
+						</xsl:when>
+						<xsl:otherwise>
+							<ctl:message>[FAILURE] The service does not provide at least one of the required protocol binding extensions.</ctl:message>
+							<ctl:fail/>
+						</xsl:otherwise>
+					</xsl:choose>
+				</xsl:when>
+				<xsl:otherwise>
+					<ctl:message>[FAILURE] Parsing or validation errors.</ctl:message>
+					<ctl:fail/>
+				</xsl:otherwise>
+			</xsl:choose>
+		</ctl:code>
+	</ctl:test>
+</ctl:package>

--- a/src/main/scripts/ctl/wcs2suite-auto.xml
+++ b/src/main/scripts/ctl/wcs2suite-auto.xml
@@ -1,4 +1,4 @@
-<ctl:suite xmlns:ctl="http://www.occamlab.com/ctl" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:wcs2="http://www.opengis.net/wcs/2.0" name="wcs2:suite">
+<ctl:suite xmlns:ctl="http://www.occamlab.com/ctl" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:wcs2="http://www.opengis.net/wcs/2.0" name="wcs2:ets-wcs20-auto">
 	<ctl:title>WCS 2.0 Test Suite</ctl:title>
 	<ctl:description>Validates WCS 2.0 Implementations</ctl:description>
 	<ctl:link>docs/wcs/2.0/</ctl:link>

--- a/src/main/scripts/ctl/wcs2suite-auto.xml
+++ b/src/main/scripts/ctl/wcs2suite-auto.xml
@@ -1,0 +1,6 @@
+<ctl:suite xmlns:ctl="http://www.occamlab.com/ctl" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:wcs2="http://www.opengis.net/wcs/2.0" name="wcs2:suite">
+	<ctl:title>WCS 2.0 Test Suite</ctl:title>
+	<ctl:description>Validates WCS 2.0 Implementations</ctl:description>
+	<ctl:link>docs/wcs/2.0/</ctl:link>
+	<ctl:starting-test>wcs2:main</ctl:starting-test>
+</ctl:suite>

--- a/src/main/scripts/ctl/wcs2suite.xml
+++ b/src/main/scripts/ctl/wcs2suite.xml
@@ -1,4 +1,4 @@
-<ctl:suite xmlns:ctl="http://www.occamlab.com/ctl" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:wcs2="http://www.opengis.net/wcs/2.0" name="wcs2:suite">
+<ctl:suite xmlns:ctl="http://www.occamlab.com/ctl" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:wcs2="http://www.opengis.net/wcs/2.0" name="wcs2:ets-wcs20">
 	<ctl:title>WCS 2.0 Test Suite</ctl:title>
 	<ctl:description>Validates WCS 2.0 Implementations</ctl:description>
 	<ctl:link>docs/wcs/2.0/</ctl:link>

